rgbd/include/rgbd/client_rgbd.h

```diff
@@ -0,0 +1,84 @@
+/**
+ * This client listens to RGBD messages on a single topic directing rgbd::Image.
+ */
+
+#ifndef RGBD_CLIENT_RGBD_H_
+#define RGBD_CLIENT_RGBD_H_
+
+#include <ros/node_handle.h>
+#include <ros/subscriber.h>
+#include <ros/callback_queue.h>
+
+#include "rgbd/types.h"
+
+#include "rgbd_msgs/RGBD.h"
+
+
+namespace rgbd {
+
+/**
+ * @brief Client which subscribes to RGBD topic
+ */
+class ClientRGBD {
+
+public:
+
+    /**
+     * @brief Constructor
+     */
+    ClientRGBD();
+
+    /**
+     * @brief Destructor
+     *
+     * image_ptr_ is not delete as the client never owns the image pointer
+     */
+    virtual ~ClientRGBD();
+
+    /**
+     * @brief intialize Initialize the client
+     * @param server_name Fully resolved server name
+     */
+    void intialize(const std::string& server_name);
+
+    /**
+     * @brief Check if the client is initialized. nextImage will not return an image if client is not initialized.
+     * @return initialized or not
+     */
+    bool initialized() { return !sub_image_.getTopic().empty(); }
+
+    /**
+     * @brief Get a new Image. If no new image has been received since the last call,
+     * no image will be written and false will be returned.
+     * @param image Image reference which will be written.
+     * @return valid image written
+     */
+    bool nextImage(Image& image);
+
+    /**
+     * @brief Get a new Image. If no new image has been received since the last call,
+     * The ImagePtr will be a nullptr
+     * @return ImagePtr to an Image or a nullptr
+     */
+    ImagePtr nextImage();
+
+protected:
+
+    ros::NodeHandle nh_;
+
+    ros::Subscriber sub_image_;
+    ros::CallbackQueue cb_queue_;
+
+    /**
+     * @brief new_image_ Track if image is updated in a callback.
+     */
+    bool new_image_;
+    Image* image_ptr_;
```

Comment:
The client never owns the image, it either returns a shared ptr, or it writes to a reference. So we set this pointer to either of these and use it internally.

Reply:
@alberth Is the use a raw pointer here correct? What is best practice?

Reply:
Shared pointers take ownership of the data block that you give them. The only entity that can construct the initial shared pointer is thus the owner of the data. Others get additional copies of the shared pointer object, and then the data block is not deleted until everybody has dropped or cleared its shared pointer.
You can easily test that yourself, make a class that prints something when its constructor and destructor is called, create an object of the class with `new`, give that pointer to a shared pointer, and let the shared pointer go out of scope. You will find the destructor gets called even though you never do a `delete` yourself.

As the image pointer is not owned by this code, you cannot wrap a shared pointer around it as you would hand over non-existing ownership of the data block.
What's worse, you would also give a reader the impression that it is a normal shared data block with all the usual semantics that go with it. In particular, the data block doesn't disappear suddenly. In this case however the owner doesn't know about the sharing, and will simply delete the data as he sees fit (hopefully after warning this class about it). A raw pointer doesn't send that message, in fact it raises questions like Janno is asking currently since the code here doesn't give any clue.

So the way out seems to me to add documentation about ownership of the data block that the pointer points at, and how safety is ensured such that the data block doesn't disappear from under our feet.

Reply:
I have added more documentation add the declaration of the `image_ptr_`

Reply:
* This is used since you cannot pass additional arguments to the callback
* You prefer to use a raw pointer to avoid unnecessary copy operations.
* If I'm not mistaken, ownership either belongs to the caller of nextImage (since this passes in a reference to an Image that's already constructed) or to the receiver of the SharedPtr

These three items are not in the docstring. @alberth would that clarify the matter in your opinion

Reply:
I update this.

Reply:
:+1: this should do